src/extra/recipes/cookiebox_deconvolve.py

```diff
@@ -0,0 +1,880 @@
+        x = prox_G(x - tau*gradT(y))
+        x_bar = x + 1*(x - x_old)
+
+        #y = prox_Fs(y + sigma*(K @ x_bar))
+        y = prox_Fs(y + sigma*grad(x_bar))
+
+        # Calculate norms
+        if k%20 == 0:
+            fidelity = np.mean(0.5*np.sqrt(np.mean((A @ x - b)**2, axis=0)), axis=-1)
+            tv = np.mean(np.sum(np.abs(grad(x)), axis=0), axis=-1)
+            energy = 1.0*fidelity + Lambda*tv
+            logging.info("[%4d] : energy %e \t fidelity %e \t TV %e" %(k,energy,fidelity,tv))
+    if is1d:
+        return x[:,0]
+
+    return np.transpose(x)
+
+def std_deconvolution(data: np.ndarray, h: np.ndarray, snr: float=5.0, n_shift: int=0):
+    r'''
+    Standard deconvolution.
+
+    Args:
+        h: Impulse response function of system to deconvolve as a 1D array.
+        data: Data to deconvolve with shape (n_samples) if a single shot is given or
+              (n_shots, n_samples) for multiple shots. It must have n_samples smaller or equal to len(h).
+        n_shift: At which sample does the impulse response of h start.
+    '''
+
+    # if data includes multiple shots, it will be (n_shots, n_samples)
+    # we need batch multplication with A, so we simply use data.T,
+    # which is shape (n_samples, n_shots), so:
+    # A is (n_samples, n_samples)
+    # x is (n_samples, n_shots)
+    # A @ x is (n_samples, n_shots)
+    # => the desired result would be trasposed
+    is1d = False
+    if len(data.shape) == 1:
+        is1d = True
+        b = data[:, None]
+    elif len(data.shape) == 2:
+        b = np.transpose(data)
+    else:
+        raise ValueError(f"Input data must either be 1D or 2D. The shape sent is {data.shape}.")
+
+    N = b.shape[0]
+
+    # make them the same size
+    if len(h) < N:
+        hl = np.concatenate((h, np.zeros(N - len(h))))
+    elif len(h) >= N:
+        hl = np.copy(h[:N])
+    hl = hl/np.sum(hl)
+
+    # roll over to match the effect of fftcovolve
+    #hl = np.roll(hl, -N//2)
+    hl = np.roll(hl, -n_shift, axis=-1)
+
+    H = np.fft.fft(hl)
+    if np.isinf(snr):
+        W = 1.0/(H+1e-20)
+    else:
+        snr2 = snr**2
+        H2 = np.abs(H)**2
+        W = snr2*np.conj(H)/(snr2*H2 + 1)
+
+    x = np.fft.ifft(W[:,None]*np.fft.fft(b, axis=0), axis=0)
+    x = np.abs(x)
+
+    if is1d:
+        return x[:,0]
+
+    return np.transpose(x)
+
+def comb(period: List[int], amplitude: List[float]):
+    c = list()
+    for T, A in zip(period, amplitude):
+        delta = np.array([1.0]+[0.0]*(T-1))
+        c += [A*delta]
+    c = np.concatenate(c)
+    return np.concatenate((np.zeros_like(c), c))
+
+class TOFResponse(SerializableMixin):
+    """
+    Given a run with the Cookiebox in counting mode, obtain its impulse response.
+
+    Example usage:
+    ```
+    tof_id = 4
+    channel = "3_A"
+    digitizer = "SQS_DIGITIZER_UTC4/ADC/1:network"
+    calib_run = open_run(proposal=900485, run=320)
+    calib_run = calib_run.select([digitizer, mpod_source], require_all=True)
```

Comment:
Should we define `mpod_source` or will that be known by the users?

Reply:
They would not and this should not be here: copy-paste mistake. Good catch!